wsidicom/instance.py

```diff
@@ -624,50 +558,264 @@ def as_tiled_full(
+        elif 'SharedFunctionalGroupsSequence' in self:
+            return self.SharedFunctionalGroupsSequence
+        return DicomSequence([])
+
+    @staticmethod
+    def _get_wsi_flavor(wsi_type: Tuple[str, str, str, str]) -> str:
+        IMAGE_FLAVOR_INDEX_IN_IMAGE_TYPE = 2
+        return wsi_type[IMAGE_FLAVOR_INDEX_IN_IMAGE_TYPE]
+
+    @staticmethod
+    def _get_spacing_between_slices_for_focal_planes(
+        focal_planes: Sequence[float]
+    ) -> float:
+        """Return spacing between slices in mm for focal planes (defined in
+        um). Spacing must be the same between all focal planes for TILED_FULL
+        arrangement.
+
+        Parameters
+        ----------
+        focal_planes: Sequence[float]
+            Focal planes to calculate spacing for.
+
+        Returns
+        ----------
+        float
+            Spacing between focal planes.
+
+        """
+        if len(focal_planes) == 1:
+            return 0.0
+        distances: Set[float] = set()
+        sorted_focal_planes = sorted(focal_planes)
+        for index in range(len(sorted_focal_planes)-1):
+            distance = (
+                sorted_focal_planes[index + 1]
+                - sorted_focal_planes[index]
+            )
+            distances.add(distance)
+        if len(distances) != 1:
+            raise NotImplementedError(
+                "Image data has non-equal spacing between slices "
+                f"{distances}, not possible to encode as TILED_FULL"
+            )
+        return distances.pop() / 1000.0
+
+    @staticmethod
+    def _get_frame_information(
+        data: OrderedDict[Tuple[str, float], 'ImageData']
+    ) -> Tuple[List[float], List[str], Size]:
+        """Return optical_paths, focal planes, and tiled size.
+        """
+        focal_planes_by_optical_path: Dict[str, List[float]] = (
+            defaultdict(list)
+        )
+        all_focal_planes: Set[float] = set()
+        tiled_sizes: Set[Size] = set()
+        for (optical_path, focal_plane), image_data in data.items():
+            focal_planes_by_optical_path[optical_path].append(focal_plane)
+            all_focal_planes.add(focal_plane)
+            tiled_sizes.add(image_data.tiled_size)
+
+        focal_planes_sparse_by_optical_path = any(
+            optical_path_focal_planes != list(all_focal_planes)
+            for optical_path_focal_planes
+            in focal_planes_by_optical_path.values()
+        )
+
+        if focal_planes_sparse_by_optical_path:
+            raise ValueError(
+                'Each optical path must have the same focal planes.'
+            )
+
+        if len(tiled_sizes) != 1:
+            raise ValueError('Expected only one tiled size')
+        tiled_size = list(tiled_sizes)[0]
+        return (
+            sorted(list(all_focal_planes)),
+            sorted(list(focal_planes_by_optical_path.keys())),
+            tiled_size
+        )
+
 
 class MetaWsiDicomFile(metaclass=ABCMeta):
```

Comment:
This doesn't look like a meta class any longer (there are no `@abstractmethod` or any other mechanism that put requirements on childclass implementations). Remove metaclass attribute and rename to `WsiDicomFileBase` to reflect that it is a base class.

Reply:
Renamed.